### PR TITLE
Add a pyro.contrib.cevae module

### PR DESCRIPTION
**Pull request by fritzo**
This refactors examples/cevae.py into a new module `pyro.contrib.cevae`. The top level interface is very simple and requires no Pyro:
```py
cevae = CEVAE(feature_dim=5)
cevae.fit(x_train, t_train, y_train)
ite = cevae.ite(x_test)  # individual treatment effect
ate = ite.mean()         # average treatment effect
```
Lower level components `Model`, `Guide`, `TraceCausalEffect_ELBO` are exposed for advanced users to customize. The `CEVAE` module can be jit traced for serving the `.ite()` method as follows:
```py
traced_cevae = cevae.jit_trace()
# ...traced_cevae can be saved and loaded...
ite = traced_cevae.ite(x_test)
```

## Reviewers

@eb8680 can you please review general interface and examples?

@neerajprad can you please review jit tracing logic?

See sphinx docs [here](http://fritzo.org/temp/pyro/html/contrib.cevae.html).

## Tested
- [x] refactoring is exercised by `pytest tests/test_examples.py -k cevae`
- [x] added a smoke test
- [x] added serialization tests
- [x] tested on a large real-world dataset (1.2M x 112)

### CODE REVIEW

File: examples/cevae.py [hunk 1]
```diff
@@ -1,11 +1,7 @@
 """
```
Comment (eb8680) [inline examples/cevae.py]: Move to `examples/contrib/cevae/`

Comment (fritzo) [reply examples/cevae.py]: done.

File: pyro/contrib/cevae/__init__.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        with pyro.plate("num_particles", num_samples, dim=-2):
+            with poutine.trace() as tr, poutine.block(hide=["y", "t"]):
+                self.guide(x)
+            with poutine.do(data=dict(t=torch.zeros(()))):
+                y0 = poutine.replay(self.model, tr.trace)(x)
+            with poutine.do(data=dict(t=torch.ones(()))):
+                y1 = poutine.replay(self.model, tr.trace)(x)
+        return (y1 - y0).mean(0)
+
+    def jit_trace(self):
```
Comment (fritzo) [inline pyro/contrib/cevae/__init__.py]: @neerajprad does this look like a reasonable interface?

Comment (neerajprad) [reply pyro/contrib/cevae/__init__.py]: This looks good. As discussed, how about naming this `to_script_module()` or `get_script_module()`?

Comment (fritzo) [reply pyro/contrib/cevae/__init__.py]: done.

**Comment by fritzo** [comment]
@eb8680 ready for review again. Main changes since last time are:
- numerical stability fixes for practical application
- include an `outcome_dist` kwarg and a few different distributions; I found `student` and `laplace` seemed to work better than `normal`. The dispatch mechanism is simple and user-extensible, relying on `.__subclasses__()`.
- more tests for serialization

File: pyro/contrib/cevae/__init__.py [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+            learning_rate=1e-3,
+            learning_rate_decay=0.1,
+            weight_decay=1e-4):
+        """
+        Train using :class:`~pyro.infer.svi.SVI` with the
+        :class:`TraceCausalEffect_ELBO` loss.
+
+        :param ~torch.Tensor x:
+        :param ~torch.Tensor t:
+        :param ~to rch.Tensor y:
```
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/contrib/cevae/__init__.py]: done

**Pull request merged**